packages/eslint-plugin-turbo/lib/utils/calculate-inputs.ts

```diff
@@ -0,0 +1,449 @@
+import crypto from "crypto";
+import fs from "fs";
+import path from "path";
+import { wildcardTests } from "../utils/wildcard-processing";
+import { dotEnv } from "../utils/dotenv-processing";
+
+import { WorkspaceConfig, getWorkspaceConfigs } from "@turbo/utils";
+import { Pipeline } from "@turbo/types";
+import { RootSchema } from "@turbo/types/src/types/config";
+
+type EnvironmentConfig = {
+  legacyConfig: string[];
+  env: string[];
+  passThroughEnv: string[] | null;
+  dotEnv: DotEnvConfig | null;
+};
+
+type EnvVar = string;
+type EnvTest = (variable: EnvVar) => boolean;
+type EnvironmentTest = {
+  legacyConfig: EnvTest;
+  env: EnvTest;
+  passThroughEnv: EnvTest;
+  dotEnv: EnvTest;
+};
+
+type DotEnvConfig = {
+  filePaths: string[];
+  hashes: {
+    [path: string]: string | null;
+  };
+};
+
+type ProjectKey = {
+  global: EnvironmentConfig;
+  globalTasks: {
+    [script: string]: EnvironmentConfig;
+  };
+  workspaceTasks: {
+    [workspace: string]: {
+      [script: string]: EnvironmentConfig;
+    };
+  };
+};
+
+type ProjectTests = {
+  global: EnvironmentTest;
+  globalTasks: {
+    [script: string]: EnvironmentTest;
+  };
+  workspaceTasks: {
+    [workspace: string]: {
+      [script: string]: EnvironmentTest;
+    };
+  };
+};
+
+// Process inputs for `EnvironmentConfig`s
+
+function processLegacyConfig(legacyConfig: string[] | undefined): string[] {
+  if (!legacyConfig) {
+    return [];
+  }
+
+  const processed = legacyConfig
+    // filter for env vars
+    .filter((dep) => dep.startsWith("$"))
+    // remove leading $
+    .map((variable) => variable.slice(1));
+
+  // After processing length is 0, 1, or more than 1.
+  switch (processed.length) {
+    case 0:
+      return [];
+    case 1:
+      return processed;
+    default:
+      return [...new Set(processed)].sort();
+  }
+}
+
+function processEnv(env: string[] | undefined): string[] {
+  if (!env) {
+    return [];
+  }
+
+  switch (env.length) {
+    case 0:
+      return [];
+    case 1:
+      return [env[0]];
+    default:
+      return [...new Set(env)].sort();
+  }
+}
+
+function processPassThroughEnv(
+  passThroughEnv: string[] | null | undefined
+): string[] | null {
+  if (!passThroughEnv) {
+    return null;
+  }
+
+  switch (passThroughEnv.length) {
+    case 0:
+      return [];
+    case 1:
+      return [passThroughEnv[0]];
+    default:
+      return [...new Set(passThroughEnv)].sort();
+  }
+}
+
+function processDotEnv(
+  workspacePath: string,
+  filePaths: string[] | null | undefined
+): DotEnvConfig | null {
+  if (!filePaths) {
+    return null;
+  }
+
+  const hashEntries: [string, string][] = [];
+  filePaths.reduce((accumulator, filePath) => {
+    const hash = crypto.createHash("sha1");
+    try {
+      const fileContents = fs.readFileSync(path.join(workspacePath, filePath));
```

Comment:
This hurts my soul, but ESLint doesn't let me have anything async at all.

Original implementation was `fs.createReadStream`.